pkg/controller/upload-controller.go

```diff
@@ -371,98 +379,118 @@ func (c *UploadController) syncHandler(key string) error {
-			}
+		var labels map[string]string
+		annotations := map[string]string{
+			AnnPodPhase: string(podPhase),
+			AnnPodReady: strconv.FormatBool(isPodReady(pod)),
+		}
+		pvc, err = updatePVC(c.client, pvc, annotations, labels)
+		if err != nil {
+			return errors.Wrapf(err, "error updating pvc %s, pod phase %s", key, podPhase)
 		}
 	}
 
 	if podSucceededFromPVC(pvc) {
-		// delete service
-		if err = c.deleteService(pvc.Namespace, resourceName); err != nil {
-			return errors.Wrapf(err, "Error deleting upload service for pvc %s", key)
-		}
-
-		dReq := podDeleteRequest{
-			namespace: pvc.Namespace,
-			podName:   resourceName,
-			podLister: c.podLister,
-			k8sClient: c.client,
-		}
-
-		// delete the pod
-		err = deletePod(dReq)
-		if err != nil {
-			return errors.Wrapf(err, "Error deleting upload pod for pvc: %s", key)
+		if err = c.cleanup(pvc); err != nil {
+			return errors.Wrap(err, "error cleaning up resources on succeeded")
 		}
 	} else {
 		// make sure the service exists
 		if _, err = c.getOrCreateUploadService(pvc, resourceName); err != nil {
-			return errors.Wrapf(err, "Error getting/creating service resource for PVC %s", key)
+			return errors.Wrapf(err, "error getting/creating service resource for PVC %s", key)
 		}
 	}
 
 	return nil
 }
 
-func (c *UploadController) getOrCreateUploadPod(pvc *v1.PersistentVolumeClaim, name string) (*v1.Pod, error) {
+func (c *UploadController) cleanup(pvc *v1.PersistentVolumeClaim) error {
+	resourceName := GetUploadResourceName(pvc.Name)
+
+	// delete service
+	err := c.deleteService(pvc.Namespace, resourceName)
+	if err != nil {
+		return errors.Wrapf(err, "error deleting upload service for pvc: %s/%s", pvc.Namespace, pvc.Name)
+	}
+
+	// delete pod
+	// we're using a req struct for now until we can normalize the controllers a bit more and share things like lister, client etc
+	// this way it's easy to stuff everything into an easy request struct, and can extend aditional behaviors if we want going forward
+	// NOTE this is a special case where the user updated annotations on the pvc to abort the upload requests, we'll add another call
+	// for this for the success case
+	dReq := podDeleteRequest{
+		namespace: pvc.Namespace,
+		podName:   resourceName,
+		podLister: c.podLister,
+		k8sClient: c.client,
+	}
+
+	err = deletePod(dReq)
+	if err != nil {
+		return errors.Wrapf(err, "error deleting upload pod for pvc: %s/%s", pvc.Namespace, pvc.Name)
+	}
+
+	return nil
+}
+
+func (c *UploadController) getOrCreateUploadPod(pvc *v1.PersistentVolumeClaim, name string, isUploadPod bool) (*v1.Pod, error) {
 	pod, err := c.podLister.Pods(pvc.Namespace).Get(name)
 
 	if k8serrors.IsNotFound(err) {
-		pod, err = CreateUploadPod(c.client, c.serverCAKeyPair, c.clientCAKeyPair.Cert, c.uploadServiceImage, c.verbose, c.pullPolicy, name, pvc, pvc.Name+"-scratch")
+		scratchPvcName := ""
+		if isUploadPod {
+			scratchPvcName = pvc.Name + "-scratch"
```

Comment:
Would it be possible to optimize the clone use case? Right now we will have 3 pvc, source/target/target-scratch. But we are sending either a tarred image or tarred files, that can be extracted directly to the target without requiring scratch space (no conversion needed).

I am fine with us just thinking about it a bit, and maybe doing a follow up PR for the optimization.

Reply:
I am sure that I am missing context here since I have not read the design document. But based on the above comment would it be correct if I say that we do not support network cloning for images that would require conversion ?

Reply:
So we are cloning one of two things:
1. An existing disk image which is already in raw format, aka it doesn't need conversion.
2. Arbitrary data that we will simply copy, aka no conversion needed.

There is no such thing as a conversion clone because the source is already in the right format.

Reply:
@awels scratch pvc is not created for clone

Reply:
Did I read the logic wrong where it is creating the the upload clone pod? I could have sworn it was creating the scratch space. If it is not, great since that is what I wanted.

Reply:
Yeah, maybe hard to tell in the diff but `isUploadPod` should be false in clone case

Reply:
Gotcha, okay well my review is completely invalid then.